<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>CWBROWSER &mdash; cwbrowser</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet" href="../../_static/css/bootstrap.min.css" media="screen" />

    
    <link rel="stylesheet" href="../../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '2.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="cwbrowser" href="../../index.html" />
  
   
       <script type="text/javascript" src="../../_static/sidebar.js"></script>
   
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <script src="../../_static/js/bootstrap.min.js" type="text/javascript"></script>

  <script src="../../_static/js/angular.min.js" type="text/javascript"></script>
  <script type="text/javascript">
    var app = angular.module("navApp", []);
    app.controller('navCtrl', ['$scope', '$window', function($scope, $window) {
      $scope.isActive = function (viewLocation) {
        var rel = $window.location.pathname.split("/")
        rel = rel[rel.length - 1]
        rel = rel.split(".")[0]
        var active = (viewLocation === rel);
        return active;};
    }]);
  </script>

  </head>
  <body role="document"><div class="banner">
  <div class="container-fluid banner-container">
    <div class="row banner-inner">
      <div class="col-xs-4">
        <img alt="Logo" src="../../_static/cwbrowser.png">
      </div>
      <div class="col-xs-8">
            <p>Python package to request CubicWeb services.</p>
      </div>
    </div>  
  </div>
</div><nav class="navbar navbar-default navbar-static-top">
  <div class="container-fluid">
    <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
        </a>
    </div>
    <div id="navbar" class="navbar-collapse collapse" ng-controller="navCtrl" ng-app="navApp">
      <ul class="nav navbar-nav">
        <li ng-class="{ active: isActive('index') }"><a href="../../index.html">Home</a></li>
        <li ng-class="{ active: isActive('installation') }"><a href="../../generated/installation.html">Installation</a></li>
        <li ng-class="{ active: isActive('gallery') }"><a href="../gallery.html">Gallery</a></li>
        <li ng-class="{ active: isActive('documentation') }"><a href="../../generated/documentation.html">API documentation</a></li>
        <!--<li class="dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false"> <span class="caret"></span></a>
          <ul class="dropdown-menu">
	        <li><a href="../../generated/cwbrowser.html">Cwbrowser</a></li>
          </ul>
        </li>-->
        <li><a href="https://github.com/AGrigis/pysphinxdoc">PYSPHINXDOC</a></li>
      </ul>
    </div>
  </div>
</nav>
<div class="content-wrapper">
    <div class="sphinxsidebar">
      <div class="sphinxsidebarwrapper">

        <!-- info setup -->
          <p class="doc-version">
           This documentation is for cwbrowser <strong>version 2.2.1</strong>
          </p>
        <p class="citing">
          If you use the software, please do not esitate to 
          <a &mdash; <a href="https://github.com/neurospin/rql_download.git">
          Report a Bug</a>.
        </p>

      <!-- toc tree -->
      <ul>
<li><a class="reference internal" href="#">Get Pathways related Genes</a><ul>
<li><a class="reference internal" href="#for-one-pathway">For one pathway</a></li>
<li><a class="reference internal" href="#for-a-list-of-pathways">For a list of pathways</a></li>
</ul>
</li>
</ul>


      </div>
    </div>
  

  <div class="content">
      
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="get-pathways-related-genes">
<span id="sphx-glr-auto-gallery-metagen-get-pathways-related-genes-py"></span><h1>Get Pathways related Genes<a class="headerlink" href="#get-pathways-related-genes" title="Permalink to this headline">¶</a></h1>
<p>Credit: A Grigis</p>
<p>In this tutorial we will send a request to a PIWS instance that contains
a genomic reference in order to get the Pathways related Genes.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Imports</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">cwbrowser</span> <span class="kn">import</span> <span class="n">CWInstanceConnection</span>
<span class="kn">from</span> <span class="nn">cwbrowser.utils</span> <span class="kn">import</span> <span class="n">ask_credential</span>

<span class="c1"># Define CubicWeb service url</span>
<span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://metagen.partenaires.cea.fr/metagen&quot;</span>
</pre></div>
</div>
<div class="section" id="for-one-pathway">
<h2>For one pathway<a class="headerlink" href="#for-one-pathway" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Define the RQL request</span>
<span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any I Where X is Pathway, X name &#39;REACTOME_DNA_REPAIR&#39;, &quot;</span>
       <span class="s2">&quot;X pathway_genes G, G hgnc_name I&quot;</span><span class="p">)</span>

<span class="c1"># Ask for login &amp; password</span>
<span class="n">login</span><span class="p">,</span> <span class="n">password</span> <span class="o">=</span> <span class="n">ask_credential</span><span class="p">()</span>

<span class="c1"># Define a connection: disable the security certificate check</span>
<span class="n">connection</span> <span class="o">=</span> <span class="n">CWInstanceConnection</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">login</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">verify</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c1"># Execute the request</span>
<span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">rset</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Number of Genes associated to &#39;REACTOME_DNA_REPAIR&#39;: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rset</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="section" id="for-a-list-of-pathways">
<h2>For a list of pathways<a class="headerlink" href="#for-a-list-of-pathways" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Define the RQL request</span>
<span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any N, I Where X is Pathway, X name IN (&#39;REACTOME_DNA_REPAIR&#39;, &quot;</span>
       <span class="s2">&quot;&#39;KEGG_RIBOSOME&#39;), X name N, X pathway_genes G, G hgnc_name I&quot;</span><span class="p">)</span>

<span class="c1"># Execute the request</span>
<span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">rset</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
<span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">rsid</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">:</span>
    <span class="n">info</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rsid</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Number of SNPs associated to &#39;REACTOME_DNA_REPAIR&#39;: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span>
    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;REACTOME_DNA_REPAIR&quot;</span><span class="p">]))</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Number of SNPs associated to &#39;KEGG_RIBOSOME&#39;: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span>
    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;KEGG_RIBOSOME&quot;</span><span class="p">]))</span>
</pre></div>
</div>
<p><strong>Total running time of the script:</strong> ( 0 minutes  0.000 seconds)</p>
<div class="sphx-glr-footer container">
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_pathways_related_genes.py" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">get_pathways_related_genes.py</span></code></a></div>
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_pathways_related_genes.ipynb" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">get_pathways_related_genes.ipynb</span></code></a></div>
</div>
<p class="sphx-glr-signature">Gallery generated by Sphinx-Gallery</p>
</div>
</div>


          </div>
        </div>
      </div>

    <div class="clearer">
    </div>
  </div>
  
</div>

<div class="footer">
    &copy; 2018, Antoine Grigis - antoine.grigis@cea.fr &lt;antoine.grigis@cea.fr&gt;.
</div>
  </body>
</html>