<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>CWBROWSER &mdash; cwbrowser</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet" href="../../_static/css/bootstrap.min.css" media="screen" />

    
    <link rel="stylesheet" href="../../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '2.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="cwbrowser" href="../../index.html" />
  
   
       <script type="text/javascript" src="../../_static/sidebar.js"></script>
   
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <script src="../../_static/js/bootstrap.min.js" type="text/javascript"></script>

  <script src="../../_static/js/angular.min.js" type="text/javascript"></script>
  <script type="text/javascript">
    var app = angular.module("navApp", []);
    app.controller('navCtrl', ['$scope', '$window', function($scope, $window) {
      $scope.isActive = function (viewLocation) {
        var rel = $window.location.pathname.split("/")
        rel = rel[rel.length - 1]
        rel = rel.split(".")[0]
        var active = (viewLocation === rel);
        return active;};
    }]);
  </script>

  </head>
  <body role="document"><div class="banner">
  <div class="container-fluid banner-container">
    <div class="row banner-inner">
      <div class="col-xs-4">
        <img alt="Logo" src="../../_static/cwbrowser.png">
      </div>
      <div class="col-xs-8">
            <p>Python package to request CubicWeb services.</p>
      </div>
    </div>  
  </div>
</div><nav class="navbar navbar-default navbar-static-top">
  <div class="container-fluid">
    <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
        </a>
    </div>
    <div id="navbar" class="navbar-collapse collapse" ng-controller="navCtrl" ng-app="navApp">
      <ul class="nav navbar-nav">
        <li ng-class="{ active: isActive('index') }"><a href="../../index.html">Home</a></li>
        <li ng-class="{ active: isActive('installation') }"><a href="../../generated/installation.html">Installation</a></li>
        <li ng-class="{ active: isActive('gallery') }"><a href="../gallery.html">Gallery</a></li>
        <li ng-class="{ active: isActive('documentation') }"><a href="../../generated/documentation.html">API documentation</a></li>
        <!--<li class="dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false"> <span class="caret"></span></a>
          <ul class="dropdown-menu">
	        <li><a href="../../generated/cwbrowser.html">Cwbrowser</a></li>
          </ul>
        </li>-->
        <li><a href="https://github.com/AGrigis/pysphinxdoc">PYSPHINXDOC</a></li>
      </ul>
    </div>
  </div>
</nav>
<div class="content-wrapper">
    <div class="sphinxsidebar">
      <div class="sphinxsidebarwrapper">

        <!-- info setup -->
          <p class="doc-version">
           This documentation is for cwbrowser <strong>version 2.2.1</strong>
          </p>
        <p class="citing">
          If you use the software, please do not esitate to 
          <a &mdash; <a href="https://github.com/neurospin/rql_download.git">
          Report a Bug</a>.
        </p>

      <!-- toc tree -->
      <ul>
<li><a class="reference internal" href="#">Get Genotype Measures</a><ul>
<li><a class="reference internal" href="#read-filter-plink">Read&amp;Filter PLINK</a></li>
<li><a class="reference internal" href="#get-some-plink-data">Get some PLINK data</a></li>
<li><a class="reference internal" href="#get-snps-of-interest">Get SNPs of interest</a></li>
<li><a class="reference internal" href="#get-the-genotype-measures-of-interest">Get the genotype measures of interest</a></li>
</ul>
</li>
</ul>


      </div>
    </div>
  

  <div class="content">
      
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="get-genotype-measures">
<span id="sphx-glr-auto-gallery-metagen-get-genotype-measures-py"></span><h1>Get Genotype Measures<a class="headerlink" href="#get-genotype-measures" title="Permalink to this headline">¶</a></h1>
<p>Credit: A Grigis</p>
<p>In this tutorial we will show you how to filter a PLINK dataset containing
genotype data. This example is an excerpt of &#8216;piws.metagen.genotype&#8217;.
Note that all the described steps have been implemented within the
web services that host your geniomic data. If the service has been
properly configured you can use the CWInstanceConnection.get_genotype_measure
method.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Imports</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">zipfile</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">cwbrowser</span> <span class="kn">import</span> <span class="n">CWInstanceConnection</span>
<span class="kn">from</span> <span class="nn">cwbrowser.utils</span> <span class="kn">import</span> <span class="n">ask_credential</span>
<span class="kn">from</span> <span class="nn">pysnptools.snpreader</span> <span class="kn">import</span> <span class="n">Bed</span>
</pre></div>
</div>
<div class="section" id="read-filter-plink">
<h2>Read&amp;Filter PLINK<a class="headerlink" href="#read-filter-plink" title="Permalink to this headline">¶</a></h2>
<p>We first define a simple function to load and filter a PLINK dataset.
This function is from &#8216;piws.metagen.genotype&#8217;.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">load_plink_bed_bim_fam_dataset</span><span class="p">(</span><span class="n">path_dataset</span><span class="p">,</span> <span class="n">snp_ids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                   <span class="n">subject_ids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">count_A1</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load a Plink bed/bim/fam dataset as a SnpData instance. Optionnally a</span>
<span class="sd">    specific list of snps or subjects can be extracted to avoid loading</span>
<span class="sd">    everything in memory.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    path_dataset: str</span>
<span class="sd">        Path to the Plink bed/bim/fam dataset, with or without .bed extension.</span>
<span class="sd">    snp_ids: list/set of str, default None</span>
<span class="sd">        Snps that should be extracted if available in the dataset.</span>
<span class="sd">        By default None, all snps are loaded.</span>
<span class="sd">    subject_ids: list of str, default None</span>
<span class="sd">        Subjects that should be extracted if available in the dataset.</span>
<span class="sd">        By default None, all subjects are loaded.</span>
<span class="sd">    count_A1: bool, default True</span>
<span class="sd">        Genotypes are provided as allele counts, A1 if True else A2.</span>

<span class="sd">    Return</span>
<span class="sd">    ------</span>
<span class="sd">    snp_data: pysnptools object</span>
<span class="sd">        PLINK data loaded by the &#39;pysnptools&#39; library.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load the metadata, without loading the genotypes</span>
    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">Bed</span><span class="p">(</span><span class="n">path_dataset</span><span class="p">,</span> <span class="n">count_A1</span><span class="o">=</span><span class="n">count_A1</span><span class="p">)</span>

    <span class="c1"># If requested, filter on snp ids</span>
    <span class="k">if</span> <span class="n">snp_ids</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">snp_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">snp_ids</span><span class="p">)</span>
        <span class="n">snp_bool_indexes</span> <span class="o">=</span> <span class="p">[(</span><span class="n">s</span> <span class="ow">in</span> <span class="n">snp_ids</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">snp_data</span><span class="o">.</span><span class="n">sid</span><span class="p">]</span>
        <span class="n">snp_data</span> <span class="o">=</span> <span class="n">snp_data</span><span class="p">[:,</span> <span class="n">snp_bool_indexes</span><span class="p">]</span>

    <span class="c1"># If requested, filter on subject ids</span>
    <span class="k">if</span> <span class="n">subject_ids</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">subject_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">subject_ids</span><span class="p">)</span>
        <span class="n">subject_bool_indexes</span> <span class="o">=</span> <span class="p">[(</span><span class="n">s</span> <span class="ow">in</span> <span class="n">subject_ids</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">snp_data</span><span class="o">.</span><span class="n">iid</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]]</span>
        <span class="n">snp_data</span> <span class="o">=</span> <span class="n">snp_data</span><span class="p">[</span><span class="n">subject_bool_indexes</span><span class="p">,</span> <span class="p">:]</span>

    <span class="c1"># Load the genotypes from the Plink dataset</span>
    <span class="n">snp_data</span> <span class="o">=</span> <span class="n">snp_data</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">snp_data</span>
</pre></div>
</div>
</div>
<div class="section" id="get-some-plink-data">
<h2>Get some PLINK data<a class="headerlink" href="#get-some-plink-data" title="Permalink to this headline">¶</a></h2>
<p>We want to download the QC Gentic PLINK data available in IMAGEN.
Define CubicWeb service url</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://imagen2.cea.fr/database&quot;</span>

<span class="c1"># Ask for login &amp; password</span>
<span class="k">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">login</span><span class="p">,</span> <span class="n">password</span> <span class="o">=</span> <span class="n">ask_credential</span><span class="p">()</span>

<span class="c1"># Define a connection</span>
<span class="n">connection</span> <span class="o">=</span> <span class="n">CWInstanceConnection</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">login</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">verify</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                                  <span class="n">server_root</span><span class="o">=</span><span class="s2">&quot;/home/{0}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">login</span><span class="p">))</span>

<span class="c1"># Define the RQL request</span>
<span class="n">rql</span> <span class="o">=</span> <span class="s2">&quot;Any G Where G is GenomicMeasure, G label &#39;QC_Genetics&#39;&quot;</span>

<span class="c1"># Create a persistent temporary directory</span>
<span class="n">tmp_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">sep</span><span class="p">,</span> <span class="s2">&quot;tmp&quot;</span><span class="p">,</span> <span class="s2">&quot;cwbrowser_data&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">tmp_dir</span><span class="p">)</span>

<span class="c1"># Execute the request</span>
<span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute_with_sync</span><span class="p">(</span><span class="n">rql</span><span class="p">,</span> <span class="n">sync_dir</span><span class="o">=</span><span class="n">tmp_dir</span><span class="p">,</span> <span class="n">nb_tries</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">rset</span><span class="p">)</span>

<span class="c1"># Unzip the downloaded file</span>
<span class="n">zip_file</span> <span class="o">=</span> <span class="n">rset</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">zip_file</span><span class="p">)</span>
<span class="n">dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">zip_file</span><span class="p">)</span>
<span class="n">fantasy_zip</span> <span class="o">=</span> <span class="n">zipfile</span><span class="o">.</span><span class="n">ZipFile</span><span class="p">(</span><span class="n">zip_file</span><span class="p">)</span>
<span class="n">fantasy_zip</span><span class="o">.</span><span class="n">extractall</span><span class="p">(</span><span class="n">dirname</span><span class="p">)</span>
<span class="n">fantasy_zip</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="n">plink_data</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="n">basename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="section" id="get-snps-of-interest">
<h2>Get SNPs of interest<a class="headerlink" href="#get-snps-of-interest" title="Permalink to this headline">¶</a></h2>
<p>We now retrieve some SNPs of interest from a list of genes</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Define the genes of interest</span>
<span class="n">gene_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;BRCA2&quot;</span><span class="p">,</span> <span class="s2">&quot;SRY&quot;</span><span class="p">]</span>

<span class="c1"># Define CubicWeb service url</span>
<span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://metagen.partenaires.cea.fr/metagen&quot;</span>

<span class="c1"># Ask for login &amp; password</span>
<span class="k">print</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">login</span><span class="p">,</span> <span class="n">password</span> <span class="o">=</span> <span class="n">ask_credential</span><span class="p">()</span>

<span class="c1"># Define a connection: disable the security certificate check</span>
<span class="n">connection</span> <span class="o">=</span> <span class="n">CWInstanceConnection</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">login</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">verify</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c1"># Define the RQL request</span>
<span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any I Where X is Gene, X hgnc_name IN ({0}), X hgnc_name N, &quot;</span>
       <span class="s2">&quot;X gene_snps S, S rs_id I&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">gene_names</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1"># Execute the request</span>
<span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">rset</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
<span class="n">snp_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Number of SNPs associated to &#39;BRCA2&#39; and &#39;TOMM22P2&#39;: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">snp_ids</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="section" id="get-the-genotype-measures-of-interest">
<h2>Get the genotype measures of interest<a class="headerlink" href="#get-the-genotype-measures-of-interest" title="Permalink to this headline">¶</a></h2>
<p>We now retrieve the genotype measures of interest from a list of SNPs.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">snp_data</span> <span class="o">=</span> <span class="n">load_plink_bed_bim_fam_dataset</span><span class="p">(</span><span class="n">plink_data</span><span class="p">,</span> <span class="n">snp_ids</span><span class="o">=</span><span class="n">snp_ids</span><span class="p">,</span>
                                          <span class="n">subject_ids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">count_A1</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">rs_ids</span> <span class="o">=</span> <span class="n">snp_data</span><span class="o">.</span><span class="n">sid</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">records</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">snp_data</span><span class="o">.</span><span class="n">iid</span><span class="p">,</span> <span class="n">snp_data</span><span class="o">.</span><span class="n">val</span><span class="p">),</span>
                            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
    <span class="n">info</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">rs_ids</span><span class="p">,</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">:]):</span>
        <span class="n">info</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;The genotype measures associated to &#39;BRCA2&#39; and &#39;TOMM22P2&#39;:&quot;</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Total running time of the script:</strong> ( 0 minutes  0.000 seconds)</p>
<div class="sphx-glr-footer container">
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_genotype_measures.py" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">get_genotype_measures.py</span></code></a></div>
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_genotype_measures.ipynb" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">get_genotype_measures.ipynb</span></code></a></div>
</div>
<p class="sphx-glr-signature">Gallery generated by Sphinx-Gallery</p>
</div>
</div>


          </div>
        </div>
      </div>

    <div class="clearer">
    </div>
  </div>
  
</div>

<div class="footer">
    &copy; 2018, Antoine Grigis - antoine.grigis@cea.fr &lt;antoine.grigis@cea.fr&gt;.
</div>
  </body>
</html>