<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
  
    <title>CWBROWSER &mdash; cwbrowser</title>
  <!-- htmltitle is before nature.css - we use this hack to load bootstrap first -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <link rel="stylesheet" href="../../_static/css/bootstrap.min.css" media="screen" />

    
    <link rel="stylesheet" href="../../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '2.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="cwbrowser" href="../../index.html" />
  
   
       <script type="text/javascript" src="../../_static/sidebar.js"></script>
   
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <script src="../../_static/js/bootstrap.min.js" type="text/javascript"></script>

  <script src="../../_static/js/angular.min.js" type="text/javascript"></script>
  <script type="text/javascript">
    var app = angular.module("navApp", []);
    app.controller('navCtrl', ['$scope', '$window', function($scope, $window) {
      $scope.isActive = function (viewLocation) {
        var rel = $window.location.pathname.split("/")
        rel = rel[rel.length - 1]
        rel = rel.split(".")[0]
        var active = (viewLocation === rel);
        return active;};
    }]);
  </script>

  </head>
  <body role="document"><div class="banner">
  <div class="container-fluid banner-container">
    <div class="row banner-inner">
      <div class="col-xs-4">
        <img alt="Logo" src="../../_static/cwbrowser.png">
      </div>
      <div class="col-xs-8">
            <p>Python package to request CubicWeb services.</p>
      </div>
    </div>  
  </div>
</div><nav class="navbar navbar-default navbar-static-top">
  <div class="container-fluid">
    <div class="navbar-header">
        <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
          <span class="sr-only">Toggle navigation</span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../index.html">
        </a>
    </div>
    <div id="navbar" class="navbar-collapse collapse" ng-controller="navCtrl" ng-app="navApp">
      <ul class="nav navbar-nav">
        <li ng-class="{ active: isActive('index') }"><a href="../../index.html">Home</a></li>
        <li ng-class="{ active: isActive('installation') }"><a href="../../generated/installation.html">Installation</a></li>
        <li ng-class="{ active: isActive('gallery') }"><a href="../gallery.html">Gallery</a></li>
        <li ng-class="{ active: isActive('documentation') }"><a href="../../generated/documentation.html">API documentation</a></li>
        <!--<li class="dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false"> <span class="caret"></span></a>
          <ul class="dropdown-menu">
	        <li><a href="../../generated/cwbrowser.html">Cwbrowser</a></li>
          </ul>
        </li>-->
        <li><a href="https://github.com/AGrigis/pysphinxdoc">PYSPHINXDOC</a></li>
      </ul>
    </div>
  </div>
</nav>
<div class="content-wrapper">
    <div class="sphinxsidebar">
      <div class="sphinxsidebarwrapper">

        <!-- info setup -->
          <p class="doc-version">
           This documentation is for cwbrowser <strong>version 2.2.1</strong>
          </p>
        <p class="citing">
          If you use the software, please do not esitate to 
          <a &mdash; <a href="https://github.com/neurospin/rql_download.git">
          Report a Bug</a>.
        </p>

      <!-- toc tree -->
      <ul>
<li><a class="reference internal" href="#">Get CPG proximity information</a><ul>
<li><a class="reference internal" href="#cpg-information">CPG information</a></li>
<li><a class="reference internal" href="#add-gene-information">Add Gene information</a></li>
</ul>
</li>
</ul>


      </div>
    </div>
  

  <div class="content">
      
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="get-cpg-proximity-information">
<span id="sphx-glr-auto-gallery-metagen-get-cpg-proximity-info-py"></span><h1>Get CPG proximity information<a class="headerlink" href="#get-cpg-proximity-information" title="Permalink to this headline">¶</a></h1>
<p>Credit: A Grigis</p>
<p>In this tutorial we will send a request to a PIWS instance that contains
a genomic reference in order to get a CPG proximity information.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Imports</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">cwbrowser</span> <span class="kn">import</span> <span class="n">CWInstanceConnection</span>
<span class="kn">from</span> <span class="nn">cwbrowser.utils</span> <span class="kn">import</span> <span class="n">ask_credential</span>

<span class="c1"># Define CubicWeb service url</span>
<span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://metagen.partenaires.cea.fr/metagen&quot;</span>
</pre></div>
</div>
<div class="section" id="cpg-information">
<h2>CPG information<a class="headerlink" href="#cpg-information" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Define the RQL request</span>
<span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any CP, CN, GN Where C is Cpg, C cg_id &#39;cg11448807&#39;, C position CP, &quot;</span>
       <span class="s2">&quot;C cpg_chromosome CR, CR name CN, C cpg_genes G, G hgnc_name GN&quot;</span><span class="p">)</span>

<span class="c1"># Ask for login &amp; password</span>
<span class="n">login</span><span class="p">,</span> <span class="n">password</span> <span class="o">=</span> <span class="n">ask_credential</span><span class="p">()</span>

<span class="c1"># Define a connection: disable the security certificate check</span>
<span class="n">connection</span> <span class="o">=</span> <span class="n">CWInstanceConnection</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">login</span><span class="p">,</span> <span class="n">password</span><span class="p">,</span> <span class="n">verify</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c1"># Execute the request</span>
<span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">rset</span><span class="p">)</span>
<span class="n">info</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cg11448807&quot;</span><span class="p">:</span> <span class="p">{}}</span>
<span class="k">for</span> <span class="n">position</span><span class="p">,</span> <span class="n">chr_name</span><span class="p">,</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">:</span>
    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;cg11448807&quot;</span><span class="p">][</span><span class="s2">&quot;position&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">position</span>
    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;cg11448807&quot;</span><span class="p">][</span><span class="s2">&quot;chr_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">chr_name</span>
    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;cg11448807&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="s2">&quot;gene_names&quot;</span><span class="p">,</span> <span class="p">{})[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Information about SNP &#39;cg11448807&#39;: &quot;</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="add-gene-information">
<h2>Add Gene information<a class="headerlink" href="#add-gene-information" title="Permalink to this headline">¶</a></h2>
<p>It is possible to perform this step with the previous one, but it is
dangerous since the gene related information we want to extract are not
always specified.
That the reson why we retrieve the gene information programmatically.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># Go through each gene</span>
<span class="k">for</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">gene_info</span> <span class="ow">in</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;cg11448807&quot;</span><span class="p">][</span><span class="s2">&quot;gene_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

    <span class="c1"># CpgIslands</span>
    <span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any I Where X is Gene, X hgnc_name &#39;{0}&#39;, X gene_cpg_islands C, &quot;</span>
           <span class="s2">&quot;C cpg_island_id I&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
    <span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
    <span class="n">gene_info</span><span class="p">[</span><span class="s2">&quot;cpg_islands&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">]</span>

    <span class="c1"># Cpgs</span>
    <span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any I Where X is Gene, X hgnc_name &#39;{0}&#39;, X gene_cpgs C, &quot;</span>
           <span class="s2">&quot;C cg_id I&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
    <span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
    <span class="n">gene_info</span><span class="p">[</span><span class="s2">&quot;cpgs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">]</span>

    <span class="c1"># Pathways</span>
    <span class="n">rql</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Any N Where X is Gene, X hgnc_name &#39;{0}&#39;, X gene_pathways P, &quot;</span>
           <span class="s2">&quot;P name N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_name</span><span class="p">))</span>
    <span class="n">rset</span> <span class="o">=</span> <span class="n">connection</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">rql</span><span class="p">)</span>
    <span class="n">gene_info</span><span class="p">[</span><span class="s2">&quot;pathways&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rset</span><span class="p">]</span>

<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Information about SNP &#39;cg11448807&#39; with gene maps: &quot;</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">(</span><span class="n">info</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Total running time of the script:</strong> ( 0 minutes  0.000 seconds)</p>
<div class="sphx-glr-footer container">
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_cpg_proximity_info.py" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">get_cpg_proximity_info.py</span></code></a></div>
<div class="sphx-glr-download container">
<a class="reference download internal" href="../../_downloads/get_cpg_proximity_info.ipynb" download=""><code class="xref download docutils literal"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">get_cpg_proximity_info.ipynb</span></code></a></div>
</div>
<p class="sphx-glr-signature">Gallery generated by Sphinx-Gallery</p>
</div>
</div>


          </div>
        </div>
      </div>

    <div class="clearer">
    </div>
  </div>
  
</div>

<div class="footer">
    &copy; 2018, Antoine Grigis - antoine.grigis@cea.fr &lt;antoine.grigis@cea.fr&gt;.
</div>
  </body>
</html>